typed/helpers.go

```diff
@@ -221,6 +221,18 @@ func keyedAssociativeListItemToPathElement(a value.Allocator, s *schema.Schema,
 	return pe, nil
 }
 
+func retainOnlyListKeys(keys []string, child value.Map) {
```

Comment:
I'm not a huge fan of this helper, but left it as is until I could get feedback on it.

The alternative I was thinking of was to add a `RetainAll` method to the `value.Map` interface that would do the filtering in place on the map. 

The issue there is that there are a ton of implementations of that Map interface and adding to all of them seemed to get a bit hairy. If we do it that way, it feels like we should be able to do better than O(mapKeys * listKeys), maybe by iterating through list keys, creating a new map with just them and changing the old map to be the new map? Kind of just thinking out loud here. 

Not sure if that would work though because I don't think you can just replace a receiver with a new map you create, right?

Anyways, rambling aside, wanted to call this out to get any thoughts you have here.